index.bs

```diff
@@ -3156,7 +3207,15 @@ The following abstract operations support the implementation of the
   1. Set |firstPendingPullInto|'s [=pull-into descriptor/buffer=] to !
      [$TransferArrayBuffer$](|firstPendingPullInto|'s [=pull-into descriptor/buffer=]).
  1. Perform ! [$ReadableByteStreamControllerInvalidateBYOBRequest$](|controller|).
- 1. If ! [$ReadableStreamHasDefaultReader$](|stream|) is true
+ 1. If |controller|.[=ReadableByteStreamController/[[pendingPullIntos]]=] is not
+    [=list/is empty|empty=],
+  1. Let |firstPendingPullInto| be
+     |controller|.[=ReadableByteStreamController/[[pendingPullIntos]]=][0].
+  1. If |firstPendingPullInto|'s [=pull-into descriptor/reader type=] is "`none`",
+   1. Perform ? [$ReadableByteStreamControllerEnqueueDetachedPullIntoToQueue$](|controller|,
+      |firstPendingPullInto|).
```

Comment:
This bit is tricky. `controller.enqueue()` may throw because we failed to allocate a new `ArrayBuffer` to hold any remaining bytes from a now-detached BYOB request:
```javascript
let controller;
const readable = new ReadableStream({
  type: 'bytes',
  start(c) { controller = c }
});
const reader1 = readable.getReader({ mode: 'byob' });

reader1.read(new Uint16Array(1));
// controller.byobRequest.view.byteLength === 2

controller.byobRequest.view[0] = 0xaa;
controller.byobRequest.respond(1);
// not enough to fill the BYOB request, so read() remains pending
// controller.byobRequest.view.byteOffset === 1
// controller.byobRequest.view.byteLength === 1

reader1.releaseLock();
// BYOB request becomes detached from request, but we cannot invalidate it yet
// since the underlying byte source might still be using it

controller.enqueue(new Uint8Array([0xbb]));
// At this point, we must move the bytes from the BYOB request ([0xaa]) into the stream's queue,
// and then push the new chunk ([0xbb]) onto the queue.
// However, moving those bytes means allocating a new ArrayBuffer containing that sub-slice,
// which might fail... and then what do we do? 🤷‍♀️
```
What should the stream do in that case?
* We must not drop any data, so the `[0xaa]` byte has to somehow end up in the queue. We do have ownership over the original `ArrayBuffer`, so it would be nice if we could just "shrink" that buffer somehow without reallocating...
* If we can't prevent data loss, should we error the stream instead? It's kinda weird though.

Reply:
We could also make `ReadableByteStreamControllerEnqueueDetachedPullIntoToQueue` *not* copy the buffer and instead do:
```javascript
ReadableByteStreamControllerEnqueueChunkToQueue(
  controller,
  firstDescriptor.buffer,
  firstDescriptor.byteOffset,
  firstDescriptor.bytesFilled
);
```
But then there's the risk that when a default reader receives this chunk, we could "leak" data before `byteOffset` or after `byteOffset + bytesFilled` which came from a previous BYOB reader. 😕

Or perhaps we could fix that leak by doing a copy inside `ReadableByteStreamControllerFillReadRequestFromQueue` instead? But then again, *that* could throw an OOM error too... 😖

Reply:
> so it would be nice if we could just "shrink" that buffer somehow without reallocating...

In theory, it feels like transferring or some variant could do this. In practice though, if we're in a situation where ArrayBuffer allocations are failing, I'm not confident we can depend on transferring to avoid such failures itself. I.e., in spec land, transfer might be shrink-in-place, but in implementations, it might need temporary buffers or some allocations or something.

So,

> If we can't prevent data loss, should we error the stream instead?

sounds reasonable to me.

In general I assume these out-of-memory cases are either (1) your computer is in trouble, let's just start erroring as soon as possible to avoid further problems; or (2) you are manipulating huge buffers, and you probably shouldn't be doing that, so let's give you an error so that you can tune your application code to do less huge buffers.